### PR TITLE
Use git history as source of truth for version genfile.

### PR DESCRIPTION
**Pull request by Keenuts**
The CHANGES file was an alternative source of truth that was trying to duplicate/replace the git history truth.

Changing the way we handle releases so we don't have to make sure our CHANGES PR is linked to the tag and tested PR, simplifying the release process.

Because we now run some git command at build-time, we need to fix kokoro's issue around dubious ownership.

Next step is to fix shaderc, which also relies on our own CHANGES file format, instead of the generated file (Any version consumer should rely on the generated file, not the sources we use).

### CODE REVIEW

File: kokoro/scripts/macos/build.sh [hunk 1]
```diff
@@ -20,6 +20,9 @@ set -e
 # Display commands being run.
 set -x
 
+# Fix dubious ownership issues on Kokoro.
+git config --global --add safe.directory /tmpfs/src/github/SPIRV-Tools
```
Comment (s-perron) [inline kokoro/scripts/macos/build.sh:24]: Could you move this below the variables below and then change the directory to `$SRC`? Having an absolute path that does not seem connected to the rest of the script could lead to problems if things change later.

Comment (s-perron) [reply kokoro/scripts/macos/build.sh:24]: Also, this seems like it should be a separate PR. It seems unrelated to the version genrule.

File: utils/update_build_version.py [hunk 2]
```diff
@@ -17,16 +17,13 @@
 # Updates an output file with version info unless the new content is the same
 # as the existing content.
 #
-# Args: <changes-file> <output-file>
+# Args: <repo-path> <output-file>
 #
 # The output file will contain a line of text consisting of two C source syntax
 # string literals separated by a comma:
-#  - The software version deduced from the given CHANGES file.
+#  - The software version deduced from the last release tag.
```
Comment (s-perron) [inline utils/update_build_version.py:24]: The `<repo-path>` parameter is no longer mentioned. It might seem useless.  You should reword this to say "from the last release tag in the repository at <repo-path>", or something like that.

File: utils/update_build_version.py [hunk 3]
```diff
@@ -63,85 +71,111 @@ def command_output(cmd, directory):
...
+
+    # This is the fallback case where git gives us no information,
+    # e.g. because the source tree might not be in a git tree.
+    # In this case, usually use a timestamp.  However, to ensure
+    # reproducible builds, allow the builder to override the wall
+    # clock time with environment variable SOURCE_DATE_EPOCH
+    # containing a (presumably) fixed timestamp.
+    timestamp = int(os.environ.get('SOURCE_DATE_EPOCH', time.time()))
+    iso_date = datetime.datetime.utcfromtimestamp(timestamp).isoformat()
+    return "unknown hash, {}".format(iso_date)
```
Comment (s-perron) [inline utils/update_build_version.py:137]: This seems to be a refactoring. This might be better in a separate PR from the actual change of behaviour. Smaller PRs make it easier to review and debug.

**Review by s-perron — CHANGES_REQUESTED** [review]
I don't have a problem with any of the changes, but I don't like them all in a single PR. It seems like you are doing many thing unrelated to "Use git history as source of truth for version genfile." That can be confusing.

If possible please split it up into multiple PRs.  Thanks.

File: utils/update_build_version.py (hunk 2 above)
Comment (Keenuts) [reply utils/update_build_version.py:24]: the repo-path is used to determine where to run the git command from, changed phrasing so it's clear.

**Review by s-perron — APPROVED** [review]

**Comment by brianosman** [comment]
Hi. I think this PR was landed incompletely? It looks like the intention was to continue supporting builds when there is no git repository present (see: `is_folder_git_repo`), but that function is never called. Instead, the script just fails when trying to get the list of tags. It would be great if the script checked for a git repo, and simply returned a standard placeholder result (like the one returned at the end of `get_description_for_head`. Would you accept that PR if I put it together?

**Comment by Keenuts** [comment]
Yes, that's a mistake, found out just now since it broke oss-fuzz run, but is_folder_git_repo should be called (and possible allow shallow clones). Preparing a PR now :)

File: BUILD.bazel [hunk 4]
```diff
@@ -104,11 +104,11 @@ py_binary(
...
 genrule(
     name = "build_version_inc",
-    srcs = ["CHANGES"],
     outs = ["build-version.inc"],
-    cmd = "SOURCE_DATE_EPOCH=0 $(location :update_build_version) $(location CHANGES) $(location build-version.inc)",
-    cmd_bat = "set SOURCE_DATE_EPOCH=0  && $(location :update_build_version) $(location CHANGES) $(location build-version.inc)",
+    cmd = "SOURCE_DATE_EPOCH=0 $(location :update_build_version) $(RULEDIR) $(location build-version.inc)",
+    cmd_bat = "set SOURCE_DATE_EPOCH=0  && $(location :update_build_version) $(RULEDIR) $(location build-version.inc)",
     exec_tools = [":update_build_version"],
+    local = True,
```
Comment (kjlubick) [inline BUILD.bazel:111]: Because local is set to true, I think `exec_tools` should be changed to `tools`. Otherwise, if we have a Mac host system that is running on a Linux RBE instance, it will try to use the Linux Python version on the Mac host.
